Fix suggestion spans for expr from macro expansions

jieyouxu opened the pull request:
### Issue #112007: rustc shows expanded `writeln!` macro in code suggestion

#### Before This PR

```
help: consider using a semicolon here
  |
6 |     };
  |      +
help: you might have meant to return this value
 --> C:\Users\hayle\.rustup\toolchains\nightly-x86_64-pc-windows-msvc\lib/rustlib/src/rust\library\core\src\macros\mod.rs:557:9
  |
55|         return $dst.write_fmt($crate::format_args_nl!($($arg)*));
  |         ++++++                                                  +
```

#### After This PR

```
help: consider using a semicolon here
   |
LL |     };
   |      +
help: you might have meant to return this value
   |
LL |         return writeln!(w, "but not here");
   |         ++++++                            +
```

### Issue #110017: `format!` `.into()` suggestion deletes the `format` macro

#### Before This PR

```
help: call `Into::into` on this expression to convert `String` into `Box<dyn std::error::Error>`
 --> /Users/eric/.rustup/toolchains/nightly-aarch64-apple-darwin/lib/rustlib/src/rust/library/alloc/src/macros.rs:121:12
  |
12|         res.into()
  |            +++++++
```

#### After This PR

```
help: call `Into::into` on this expression to convert `String` into `Box<dyn std::error::Error>`
   |
LL |     Err(format!("error: {x}").into())
   |                              +++++++
```

---

Fixes #112007.
Fixes #110017.

jieyouxu committed:
Fix suggestions span for when expr is from macro expansions

cjgillot commented on compiler/rustc_hir_typeck/src/fn_ctxt/suggestions.rs:
```diff
@@ -907,11 +907,20 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
+                }
+                let parent = self.tcx.hir().get_enclosing_scope(node_id).unwrap_or(node_id);
+                let parent_span = self.tcx.hir().span(parent);
+                let span = expr.span.find_ancestor_inside(parent_span).unwrap_or(expr.span);
```
I wonder if we should have a `Span::find_suggestable_ancestor` which walks the expansion tree up, until we are not in an external macro any more. With the definition of external macros given by `rustc_middle::lint::in_external_macro`.

At least, the two instances of this code deserve a helper function somewhere.

jieyouxu replied:
> At least, the two instances of this code deserve a helper function somewhere.

I extracted this code into a helper function `find_suggestable_ancestor_span` inside `rustc_hir_typeck/src/fn_ctxt/suggestions.rs` to avoid the duplication.

cjgillot commented on compiler/rustc_hir_typeck/src/fn_ctxt/suggestions.rs:
```diff
@@ -907,11 +907,20 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
             };
             let ty = self.normalize(expr.span, ty);
             if self.can_coerce(found, ty) {
+                let mut node_id = self.tcx.hir().parent_id(expr.hir_id);
```
Could you write an explanation for the logic here? I'm not sure I fully understand the stopping condition here.

jieyouxu replied:
The intent is (I'm not 100% sure if this logic is correct) that we keep peeling away parents of the expression until we are no longer in a macro expansion (there might be multiple macro expansions in a chain, e.g. `outer!()` expands to `inner!()` etc.). Once we find a parent who is no longer from a macro expansion, we use that as a stopping condition for `find_ancestor_inside` to find a span that's contained within the span of the parent.

jieyouxu replied:
If I don't do this "peeling away parents" business, it does not correctly work for `tests/ui/typeck/issue-110017-format-into-help-deletes-macro.rs` with just something like

```rust
let parent = self.tcx.hir().get_enclosing_scope(expr.hir_id).unwrap_or(expr.hir_id);
let parent_span = self.tcx.hir().span(parent);
let span = expr.span.find_ancestor_inside(parent_span).unwrap_or(expr.span);
```

```
error[E0308]: mismatched types
 --> tests/ui/typeck/issue-110017-format-into-help-deletes-macro.rs:2:9
  |
2 |     Err(format!("error: {x}"))
  |         ^^^^^^^^^^^^^^^^^^^^^ expected `Box<dyn Error>`, found `String`
  |
  = note: expected struct `Box<dyn std::error::Error>`
             found struct `String`
  = note: this error originates in the macro `format` (in Nightly builds, run with -Z macro-backtrace for more info)
help: call `Into::into` on this expression to convert `String` into `Box<dyn std::error::Error>`
 --> /home/jieyouxu/repos/rust/library/alloc/src/macros.rs:121:12
  |
12|         res.into()
  |            +++++++

error[E0308]: mismatched types
  --> tests/ui/typeck/issue-110017-format-into-help-deletes-macro.rs:20:9
   |
20 |     Err(outer!(x))
   |         ^^^^^^^^^ expected `Box<dyn Error>`, found `String`
   |
   = note: expected struct `Box<dyn std::error::Error>`
              found struct `String`
   = note: this error originates in the macro `format` which comes from the expansion of the macro `outer` (in Nightly builds, run with -Z macro-backtrace for more info)
help: call `Into::into` on this expression to convert `String` into `Box<dyn std::error::Error>`
  --> /home/jieyouxu/repos/rust/library/alloc/src/macros.rs:121:12
   |
121|         res.into()
   |            +++++++

```

cjgillot commented on compiler/rustc_hir_typeck/src/fn_ctxt/suggestions.rs:
```diff
@@ -907,18 +907,34 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
+        let parent_span = self.tcx.hir().span(parent);
+        let span = expr.span.find_ancestor_inside(parent_span).unwrap_or(expr.span);
+        span
+    }
```
What happens if the function itself is generated by a macro?
What happens if a local macro generates non-trivial code? Should we suggest inside that macro?
Should we stop the parent walk when we stop seeing expressions?

To be clear: I think this heuristic is too strong, and should be refined a bit.

jieyouxu replied:
> What happens if the function itself is generated by a macro?

For something like

```rust
macro_rules! entire_fn {
    () => {
        pub fn baz(x: &str) -> Result<(), Box<dyn std::error::Error>> {
            Err(format!("error: {x}"))
            //~^ ERROR mismatched types
        }
    }
}

entire_fn!();
```

I could get the span to be inside `entire_fn` via `find_ancestor_in_same_ctxt` instead of `find_ancestor_inside`.

> What happens if a local macro generates non-trivial code? Should we suggest inside that macro?

I'm having a hard time trying to come up with a test example for this, do you have an example in mind?

> Should we stop the parent walk when we stop seeing expressions?

I'm not quite sure about this... I don't think it would be right if we stopped the parent walk when we stop seeing expressions? The parent could very well be items and not expressions?

cjgillot replied:
> I'm having a hard time trying to come up with a test example for this, do you have an example in mind?

In your test, having the macro expand to `Err(format!(...))`, instead of just the `format!`.

jieyouxu replied:
Yeah unfortunately I don't think this can work...

Given something like

```rust
macro_rules! nontrivial {
    ($x: expr) => {
        Err(format!("error: {}", $x))
        //~^ ERROR mismatched types
    }
}

pub fn qux(x: &str) -> Result<(), Box<dyn std::error::Error>> {
    nontrivial!(x)
}
```

my current code will suggest `nontrivial!(x).into()` which isn't correct. Should we just suppress `Into::into` suggestion when `expr`'s span comes from a macro expansion...?

davidtwco replied:
I think it's okay to continue making a suggestion for the macro code, perhaps only if it's a locally-defined macro (if that's easy to check here). I don't see much value in making that suggestion if the macro is defined in an external crate.

jieyouxu commented:
### Changes Since Last Force-Push

Whereas previously used incorrect parent node searching logic (I meant to have used *parent callsites* instead), the new changes differentiates treatment of finding a suitable span for suggesting `return expr;` vs `expr.into()` when *local* macros are involved.

For suggesting `.into()`, we now recursively try to look for the ancestor span which *shares the same context* as the initial `expr.span`. This will recursively look into *local macros* until the span inside the ancestor local macro is found. It will stop recursing as soon as the context of a potential parent callsite changes, such as if the potential parent callsite is in a foreign macro.

For suggesting `return expr;`, we use `Span::find_ancestor_inside` to find an ancestor for `expr.span` which is inside the owner node of the function. This will be less aggressive than the `.into()` suggestion, and avoids looking into local macros, but instead tries to stay as close to the function as possible.

davidtwco reviewed (approved):
This seems good to me now, but I'm not 100% confident there aren't cases which wouldn't work. Does this resolve your previous concerns, @cjgillot?

cjgillot commented:
Thanks.
Let's go with this.
@bors r+

bors commented:
:pushpin: Commit 25d065b2d476053102b05bbd5c4650a86ddc2c94 has been approved by `cjgillot`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=cjgillot 25d065b2d476053102b05bbd5c4650a86ddc2c94 -->
<!-- homu: {"type":"Approved","sha":"25d065b2d476053102b05bbd5c4650a86ddc2c94","approver":"cjgillot","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 25d065b2d476053102b05bbd5c4650a86ddc2c94 with merge c115ec11d2087050dc12c5c83959979aa98bb3e5...
<!-- homu: {"type":"BuildStarted","head_sha":"25d065b2d476053102b05bbd5c4650a86ddc2c94","merge_sha":"c115ec11d2087050dc12c5c83959979aa98bb3e5"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/5749321417/job/15586840319)
Approved by: cjgillot
Pushing c115ec11d2087050dc12c5c83959979aa98bb3e5 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"cjgillot","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/5749321417/job/15586840319"},"merge_sha":"c115ec11d2087050dc12c5c83959979aa98bb3e5"} -->

rust-timer commented:
Finished benchmarking commit (c115ec11d2087050dc12c5c83959979aa98bb3e5): [comparison URL](https://perf.rust-lang.org/compare.html?start=a922d1c0da0c904ffe9de256d926180613f1ed81&end=c115ec11d2087050dc12c5c83959979aa98bb3e5&stat=instructions:u).

### Overall result: no relevant changes - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=a922d1c0da0c904ffe9de256d926180613f1ed81&end=c115ec11d2087050dc12c5c83959979aa98bb3e5&stat=instructions:u)
This benchmark run did not return any relevant results for this metric.

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=a922d1c0da0c904ffe9de256d926180613f1ed81&end=c115ec11d2087050dc12c5c83959979aa98bb3e5&stat=max-rss)
This benchmark run did not return any relevant results for this metric.

### [Cycles](https://perf.rust-lang.org/compare.html?start=a922d1c0da0c904ffe9de256d926180613f1ed81&end=c115ec11d2087050dc12c5c83959979aa98bb3e5&stat=cycles:u)
This benchmark run did not return any relevant results for this metric.

### [Binary size](https://perf.rust-lang.org/compare.html?start=a922d1c0da0c904ffe9de256d926180613f1ed81&end=c115ec11d2087050dc12c5c83959979aa98bb3e5&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 649.583s -> 649.925s (0.05%)

ehuss commented:
@jieyouxu The test added here is causing some issues with updating `rustfix`. Newer versions of rustfix are better at applying certain kinds of suggestions, but unfortunately the `ui/typeck/issue-112007-leaked-writeln-macro-internals.rs` test still has a suggestion that tries to modify an external crate (libcore).

Do you think you would be able to help look at trying to fix that? 

You can reproduce by changing `src/tools/compiletest/Cargo.toml` to have `rustfix = "0.8.1"`.

The problem is that the diagnostic has a suggestion to add a `?`:

```
use the `?` operator to extract the `Result<(), std::fmt::Error>` value, propagating a `Result::Err` value to the caller
```

and it points to inserting the `?` into `library/core/src/macros/mod.rs`. I'm guessing that particular suggestion also needs to do a similar dance as what this PR does of peeling back the span to point to the user's code?

jieyouxu commented:
> The test added here is causing some issues with updating `rustfix`. Newer versions of rustfix are better at applying certain kinds of suggestions, but unfortunately the `ui/typeck/issue-112007-leaked-writeln-macro-internals.rs` test still has a suggestion that tries to modify an external crate (libcore).

That's... unfortunate. I'll take a look.

jieyouxu commented:
Am I understanding correctly that because rustfix got better at applying certain suggestions, it exposed another case of pointing into library code?

ehuss commented:
> Am I understanding correctly that because rustfix got better at applying certain suggestions, it exposed another case of pointing into library code?

Yea.

ehuss commented:
Hm, thinking about it some more, just fixing the span of the `?` suggestion won't work, since the test won't be able to apply both the `return…;` and `?` suggestions at the same time (they conflict with one another).

I'm not sure how to fix that.

jieyouxu commented:
That is *extremely* unfortunate. How would rustfix/rustc even handle conflict resolution here? It's not immediately obvious to me which suggestion is better...

ehuss commented:
It's not uncommon for rustc to suggest conflicting suggestions, which is fine as long as they are not marked MachineApplicable. Sometimes there really are several legitimate options, and it isn't feasible for rustc to know which one is the correct one. The user has to decide.

On the `rustfix` side, it currently only handles MachineApplicable suggestions in `cargo fix`, so we haven't worried about this too much. In the future we want to add an interactive mode to `cargo fix` to allow the user to choose from multiple options when there is a choice (of MaybeIncorrect choices), but it would have to be a manual choice.

UI tests are a little unusual in that they apply all applicabilities by default, even if they are "maybe incorrect" or "has placeholders". For example, `issue-112007-leaked-writeln-macro-internals.rs` has both MaybeIncorrect and HasPlaceholders suggestions, which are ignored by `cargo fix`.

jieyouxu commented:
Wait, I forgot this wasn't even machine applicable suggestions, both are MaybeIncorrect

jieyouxu commented:
Welp, one step at a time. I'll try to fix the `?` pointing into library issue first. Still not sure how to fix overlapping MaybeIncorrect suggestions. I wonder if we have to deal with this on the compiletest test to somehow determine that if we have overlapping MaybeIncorrect suggestions, that we try to synthesize multiple `.fixed` files and verify each of them individually compiles...?

ehuss commented:
BTW, thanks for responding so fast and taking a look! 😃

jieyouxu commented:
> just fixing the span of the `?` suggestion won't work

well the fix for `?` suggestion span seems rather easy, it's exactly the same logic as the `(...).into()` span in this PR...


```diff
diff --git a/compiler/rustc_hir_typeck/src/fn_ctxt/suggestions.rs b/compiler/rustc_hir_typeck/src/fn_ctxt/suggestions.rs
index f09af999957..d349da95266 100644
--- a/compiler/rustc_hir_typeck/src/fn_ctxt/suggestions.rs
+++ b/compiler/rustc_hir_typeck/src/fn_ctxt/suggestions.rs
@@ -1927,8 +1927,14 @@ pub(crate) fn suggest_missing_unwrap_expect(
             None => sugg.to_string(),
         };

+        let mut span = expr.span;
+        while expr.span.eq_ctxt(span) && let Some(parent_callsite) = span.parent_callsite()
+        {
+            span = parent_callsite;
+        }
+
         err.span_suggestion_verbose(
-            expr.span.shrink_to_hi(),
+            span.shrink_to_hi(),
             msg,
             sugg,
             Applicability::HasPlaceholders,

```

jieyouxu commented:
Actually, we should just do the simplest and most correct thing here:

**delete `//@ run-rustfix`**

This will definitely fix the issue. And since the suggestions are MaybeIncorrect anyway, it's okay if they conflict and don't cleanly apply together, we can verify that they don't leak stdlib implementation details already via stderr, we don't *need* .fixed!

compiler-errors commented on compiler/rustc_hir_typeck/src/fn_ctxt/suggestions.rs:
```diff
@@ -1177,10 +1182,16 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                 ),
             ))
         {
+            let mut span = expr.span;
+            while expr.span.eq_ctxt(span) && let Some(parent_callsite) = span.parent_callsite()
+            {
+                span = parent_callsite;
+            }
```
Why is this different than [`Span::find_ancestor_in_same_ctxt`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_span/span_encoding/struct.Span.html#method.find_ancestor_in_same_ctxt)?
